### PR TITLE
Wire splices/node:edits back to cmdr

### PR DESCRIPTION
**Pull request by jnwatson**
Also, Fix issue where log writes/events/provenance were writing even if no new data

### CODE REVIEW

File: synapse/cmds/cortex.py [hunk 1]
```diff
@@ -210,6 +217,13 @@ class StormCmd(s_cli.Cmd):
         --hide-tags: Do not print tags.
         --hide-props: Do not print secondary properties.
         --hide-unknown: Do not print messages which do not have known handlers.
+        --show-nodeedits:  Show full nodeedits (otherwise printed as a single .)
+        --editformat <format>: What format of edits the server shall emit.
+                Options are
+                   * nodeedits (default),
+                   * splices (similar to < 0.2.0),
+                   * count (just counts of nodeedits), or
```
Comment (invisig0th) [inline synapse/cmds/cortex.py]: i dont think this needs comma/or when it's a bullet list

File: synapse/lib/view.py [hunk 2]
```diff
@@ -227,13 +236,36 @@ def pack(self):
...
+            if kind == 'node:edits':
+                if editformat == 'nodeedits':
+                    yield mesg
+                    continue
+
+                if editformat == 'none':
+                    continue
+
+                if editformat == 'count':
+                    necount += 1
```
Comment (invisig0th) [inline synapse/lib/view.py]: maybe count should be the count of the total node edits to reflect the work the system is doing?

File: synapse/lib/view.py [hunk 3]
```diff
@@ -227,13 +236,36 @@ def pack(self):
...
         await synt.worker(runStorm())
 
+        necount = 0
+        editformat = opts.get('editformat', 'nodeedits')
+
         while True:
 
             mesg = await chan.get()
+            kind = mesg[0]
+
```
Comment (invisig0th) [inline synapse/lib/view.py]: Would we want to maybe short circuit "node" type messages here to speed them along?

File: synapse/cmds/cortex.py (hunk 1 above)
Comment (jnwatson) [reply synapse/cmds/cortex.py]: It does if you're writing in prose.

File: synapse/lib/view.py (hunk 2 above)
Comment (jnwatson) [reply synapse/lib/view.py]: How would that be different from what we're currently doing?

File: synapse/lib/view.py (hunk 3 above)
Comment (jnwatson) [reply synapse/lib/view.py]: ok

File: synapse/lib/view.py (hunk 2 above)
Comment (Cisphyx) [reply synapse/lib/view.py]: I think it means increment by the number of edits in the nodeedits

Comment (invisig0th) [reply synapse/lib/view.py]: Yep!  The total number of edits within the nodeedits.

File: synapse/cmds/cortex.py [hunk 4]
```diff
@@ -35,8 +40,8 @@ class Log(s_cli.Cmd):
     # Disable logging and close the current file
     log --off
 
-    # Enable logging, but only log splices. Log them as jsonl instead of mpk.
-    log --on --splices-only --format jsonl
+    # Enable logging, but only log edits. Log them as jsonl instead of mpk.
+    log --on --edits-only --format jsonl
```
Comment (vEpiphyte) [inline synapse/cmds/cortex.py]: This fails to run.

```
cli> log --format jsonl --on --path netest.jsonl                                                                                                                                         
Starting logfile at [netest.jsonl]
cli> storm [inet:ipv4=1.2.3.4 :asn=2020]                                                                                                                                                 
Executing query at 2020/03/12 14:53:25.375
.2020-03-12 14:53:25,398 [ERROR] Failed to serialize message: [Object of type bytes is not JSON serializable] [cortex.py:save:Thread-164:MainProcess]
2020-03-12 14:53:25,398 [ERROR] Failed to serialize message: [Object of type bytes is not JSON serializable] [cortex.py:save:Thread-164:MainProcess]
.inet:ipv4=1.2.3.4
        .created = 2020/03/12 14:53:25.389
        :asn = 2020
        :type = unicast
complete. 1 nodes in 22 ms (45/sec).
```

Comment (vEpiphyte) [reply synapse/cmds/cortex.py]:
```$ cat netest.jsonl 
["init", {"task": "dd40c08c7430ffaaa94d7e4931fa9ef2", "text": "[inet:ipv4=1.2.3.4 :asn=2020]", "tick": 1584024805375}]
["node", [["inet:ipv4", 16909060], {"iden": "20153b758f9d5eaaa38e4f4a65c36da797c3e59e549620fa7c4895e1a920991f", "path": {}, "props": {".created": 1584024805389, "asn": 2020, "type": "unicast"}, "repr": "1.2.3.4", "reprs": {".created": "2020/03/12 14:53:25.389", "asn": "2020"}, "tagpropreprs": {}, "tagprops": {}, "tags": {}}]]
["fini", {"count": 1, "tock": 1584024805397, "took": 22}]
```

Comment (jnwatson) [reply synapse/cmds/cortex.py]: good catch. Fixed.

File: synapse/cmds/cortex.py [hunk 5]
```diff
@@ -248,13 +264,26 @@ def __init__(self, cli, **opts):
             'fini': self._onFini,
             'print': self._onPrint,
             'warn': self._onWarn,
-            'err': self._onErr
+            'err': self._onErr,
+            'node:edits': self._onNodeEdits,
+            'node:edits:count': self._onNodeEditsCount
```
Comment (vEpiphyte) [inline synapse/cmds/cortex.py]: trailling `,` please

File: synapse/cmds/cortex.py [hunk 6]
```diff
@@ -391,35 +424,29 @@ def _onErr(self, mesg):
...
+                else:
+                    func(mesg, opts)
 
         except s_exc.SynErr as e:
 
             if e.errinfo.get('errx') == 'CancelledError':
                 self.printf('query canceled.')
                 return
 
-            raise
```
Comment (vEpiphyte) [inline synapse/cmds/cortex.py]: why remove this?

File: synapse/common.py [hunk 7]
```diff
@@ -103,7 +103,7 @@ def buid(valu=None):
 
 def ehex(byts):
     '''
-    Encode a set of bytes to a string using binascii.hexlify.
+    Encode a bytes to a string using binascii.hexlify.
```
Comment (vEpiphyte) [inline synapse/common.py]: ``Encode a bytes`` sounds odd.

File: synapse/common.py [hunk 8]
```diff
@@ -648,3 +648,35 @@ def config(conf, confdefs):
...
+        newedit = (ehex(nodeedit[0]), *nodeedit[1:])
+        retn.append(newedit)
+
+    return retn
+
+def unjsonsafe_nodeedits(nodeedits):
+    retn = []
+    for nodeedit in nodeedits:
+        buid = nodeedit[0]
+        if isinstance(buid[0], str):
```
Comment (vEpiphyte) [inline synapse/common.py]: why do the index into the object to check the instance?  seems like extra work vs just checking ``buid`` directly.

File: synapse/common.py [hunk 9]
```diff
@@ -648,3 +648,35 @@ def config(conf, confdefs):
...
+            newedit = (uhex(nodeedit[0]), *nodeedit[1:])
```
Comment (vEpiphyte) [inline synapse/common.py]: why not use ``buid`` when you've already set it to the value of ``nodeedit[0]`` ?

File: synapse/cortex.py [hunk 10]
```diff
@@ -2910,15 +2911,16 @@ def getCoreInfo(self):
 
     async def _initCoreMods(self):
 
-        for ctor, modu in list(self.modules.items()):
+        with s_provenance.claim('init', meth='_initCoreMods'):
```
Comment (vEpiphyte) [inline synapse/cortex.py]: :+1:

File: synapse/cmds/cortex.py (hunk 6 above)
Comment (jnwatson) [reply synapse/cmds/cortex.py]: oops

File: synapse/lib/layer.py [hunk 11]
```diff
@@ -1193,21 +1195,31 @@ def getNodeTag(self, buid, tag):
...
-        return changes
+        return results
 
     async def _storNodeEdit(self, nodeedit):
         '''
         Execute a series of storage operations for the given node.
 
-        Returns a (buid, edits) tuple of the actual changes.
+        Args:
+            node
```
Comment (vEpiphyte) [inline synapse/lib/layer.py]: wrong argname

File: synapse/lib/types.py [hunk 12]
```diff
@@ -1233,6 +1233,7 @@ class Data(Type):
     stortype = s_layer.STOR_TYPE_MSGP
 
     def norm(self, valu):
+        s_common.reqjsonsafe(valu)
```
Comment (vEpiphyte) [inline synapse/lib/types.py]: This should be caugh and raise a BadTypeValu from the raised exception

File: synapse/tests/test_cortex.py [hunk 13]
```diff
@@ -2675,6 +2675,10 @@ class Bawk:
                 nodelist1 = [node.pack() for node in nodelist1]
                 self.eq(nodelist0, nodelist1)
 
+                # Try a nodeedits we might get from cmdr
+                cmdrnodeedits = s_common.jsonsafe_nodeedits(editlist[0])
+                await prox1.addFeedData('syn.nodeedits', [cmdrnodeedits])
```
Comment (vEpiphyte) [inline synapse/tests/test_cortex.py]: Is there an action from this edit we could validate as having been made to the cortex?

File: synapse/tests/test_lib_httpapi.py [hunk 14]
```diff
@@ -681,6 +681,7 @@ class ReqAuthHandler(s_httpapi.Handler):
                         if mesg[0] == 'node':
                             node = mesg[1]
 
+                    self.nn(node)
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_httpapi.py]: ?

File: synapse/tests/test_model_base.py [hunk 15]
```diff
@@ -77,6 +77,10 @@ class BaseTest(s_t_utils.SynTest):
                 self.eq(node.get('data'), ('some', 'data', 'here'))
                 self.eq(node.get('name'), 'Magic Pony')
 
+                # Raise on non-json-safe values
+                props['data'] = {(1, 2): 'foo'}
+                await self.asyncraises(s_exc.MustBeJsonSafe, snap.addNode('graph:event', iden, props=props))
```
Comment (vEpiphyte) [inline synapse/tests/test_model_base.py]: Do we have an explicit failure for bytes as well?

File: synapse/tests/test_lib_httpapi.py (hunk 14 above)
Comment (jnwatson) [reply synapse/tests/test_lib_httpapi.py]: that's valid

File: synapse/tests/test_model_base.py (hunk 15 above)
Comment (jnwatson) [reply synapse/tests/test_model_base.py]: ok

File: synapse/lib/types.py [hunk 16]
```diff
@@ -1233,7 +1232,10 @@ class Data(Type):
     stortype = s_layer.STOR_TYPE_MSGP
 
     def norm(self, valu):
-        s_common.reqjsonsafe(valu)
+        try:
+            s_common.reqjsonsafe(valu)
+        except s_exc.MustBeJsonSafe as e:
+            raise s_exc.BadTypeValu() from e
```
Comment (vEpiphyte) [inline synapse/lib/types.py]: mesg = ... ?

Comment (vEpiphyte) [reply synapse/lib/types.py]: and type=self.type.full ? we have some common bits that help people understand what form failed to norm a thing.

Comment (vEpiphyte) [reply synapse/lib/types.py]: ``name=self.name``

**Pull request merged**